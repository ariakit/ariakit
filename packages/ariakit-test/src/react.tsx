import type { ReactElement } from "react";
import { StrictMode } from "react";
import * as ReactTestingLibrary from "@testing-library/react";
import { flushMicrotasks, nextFrame, wrapAsync } from "./__utils.js";

export * from "./index.js";

export interface RenderOptions
  extends Omit<ReactTestingLibrary.RenderOptions, "queries"> {
  strictMode?: boolean;
}

export async function render(ui: ReactElement, options?: RenderOptions) {
  const wrapper = (props: { children: ReactElement }) => {
    const Wrapper = options?.wrapper;
    const element = Wrapper ? <Wrapper {...props} /> : props.children;
    if (!options?.strictMode) return element;
    return <StrictMode>{element}</StrictMode>;
  };

  return wrapAsync(async () => {
    const { unmount } = ReactTestingLibrary.render(ui, { ...options, wrapper });
    await flushMicrotasks();
    await nextFrame();
    await flushMicrotasks();
    return unmount;
  });
}
