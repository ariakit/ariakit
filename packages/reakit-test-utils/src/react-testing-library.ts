export {
  act,
  fireEvent,
  render,
  wait,
  waitForDomChange,
  waitForElement,
  waitForElementToBeRemoved
} from "@testing-library/react";
