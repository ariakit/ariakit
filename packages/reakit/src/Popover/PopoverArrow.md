```jsx
import { Popover } from "reakit";

<Popover.Arrow />
```
