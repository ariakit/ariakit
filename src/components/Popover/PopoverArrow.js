import styled from "styled-components";
import { prop } from "styled-tools";
import as from "../../enhancers/as";
import Arrow from "../Arrow";
import Box from "../Box";

const ArrowBox = Arrow.as(Box);

const PopoverArrow = styled(ArrowBox)`
  position: absolute;
  color: white;
  border: inherit;
  border-top: 0;
  font-size: 1.25em;
  border-radius: 0;
  [data-placement^="top"] > & {
    top: 100%;
  }
  [data-placement^="right"] > & {
    right: 100%;
    transform: rotateZ(90deg);
  }
  [data-placement^="bottom"] > & {
    bottom: 100%;
    transform: rotateZ(180deg);
  }
  [data-placement^="left"] > & {
    left: 100%;
    transform: rotateZ(-90deg);
  }
  ${prop("theme.PopoverArrow")};
`;

export default as("div")(PopoverArrow);
