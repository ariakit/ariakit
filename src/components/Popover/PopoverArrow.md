```jsx
import { InlineBlock } from "reakit";

<InlineBlock relative>
  <Popover.Arrow />
</InlineBlock>
```