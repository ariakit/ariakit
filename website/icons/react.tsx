import type { SVGAttributes } from "react";

export function React(props: SVGAttributes<SVGSVGElement>) {
  return (
    <svg viewBox="-11.5 -10.23174 23 20.46348" {...props}>
      <circle
        className="fill-blue-600 dark:fill-[#61dafb]"
        cx="0"
        cy="0"
        r="2.05"
      />
      <g
        className="stroke-blue-600 dark:stroke-[#61dafb]"
        strokeWidth="1"
        fill="none"
      >
        <ellipse rx="11" ry="4.2" />
        <ellipse rx="11" ry="4.2" transform="rotate(60)" />
        <ellipse rx="11" ry="4.2" transform="rotate(120)" />
      </g>
    </svg>
  );
}
