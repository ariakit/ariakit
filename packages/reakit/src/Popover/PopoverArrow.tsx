import * as React from "react";
import { createComponent } from "reakit-system/createComponent";
import { createHook } from "reakit-system/createHook";
import { useForkRef } from "reakit-utils/useForkRef";
import { BoxOptions, BoxHTMLProps, useBox } from "../Box/Box";
import { PopoverStateReturn } from "./PopoverState";
import { POPOVER_ARROW_KEYS } from "./__keys";

export type PopoverArrowOptions = BoxOptions &
  Pick<
    Partial<PopoverStateReturn>,
    "unstable_arrowRef" | "unstable_arrowStyles"
  > &
  Pick<PopoverStateReturn, "placement"> & {
    /** Arrow's size */
    size?: number | string;
  };

export type PopoverArrowHTMLProps = BoxHTMLProps;

export type PopoverArrowProps = PopoverArrowOptions & PopoverArrowHTMLProps;

export const usePopoverArrow = createHook<
  PopoverArrowOptions,
  PopoverArrowHTMLProps
>({
  name: "PopoverArrow",
  compose: useBox,
  keys: POPOVER_ARROW_KEYS,

  useOptions({ size = 30, ...options }) {
    return { size, ...options };
  },

  useProps(options, { ref: htmlRef, style: htmlStyle, ...htmlProps }) {
    const [placement] = options.placement.split("-");
    const transformMap: Record<string, string> = {
      top: "rotateZ(180deg)",
      right: "rotateZ(-90deg)",
      bottom: "rotateZ(360deg)",
      left: "rotateZ(90deg)",
    };
    const { unstable_arrowStyles: arrowStyles } = options;
    const transform = transformMap[placement];

    const children = React.useMemo(
      () => (
        <svg viewBox="0 0 30 30" style={{ transform }}>
          <path
            className="stroke"
            d="M23.7,27.1L17,19.9C16.5,19.3,15.8,19,15,19s-1.6,0.3-2.1,0.9l-6.6,7.2C5.3,28.1,3.4,29,2,29h26
        C26.7,29,24.6,28.1,23.7,27.1z"
          />
          <path
            className="fill"
            d="M23,27.8c1.1,1.2,3.4,2.2,5,2.2h2H0h2c1.7,0,3.9-1,5-2.2l6.6-7.2c0.7-0.8,2-0.8,2.7,0L23,27.8L23,27.8z"
          />
        </svg>
      ),
      [transform]
    );

    return {
      ref: useForkRef(options.unstable_arrowRef, htmlRef),
      style: {
        ...arrowStyles,
        fontSize: options.size,
        width: "1em",
        height: "1em",
        pointerEvents: "none",
        [placement]: "100%",
        ...htmlStyle,
      },
      children,
      ...htmlProps,
    };
  },
});

export const PopoverArrow = createComponent({
  as: "div",
  memo: true,
  useHook: usePopoverArrow,
});
