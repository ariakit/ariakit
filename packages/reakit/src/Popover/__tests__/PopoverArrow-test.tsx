import * as React from "react";
import { render } from "react-testing-library";
import PopoverArrow from "../PopoverArrow";

test("html attrs", () => {
  const { getByTestId } = render(
    <PopoverArrow data-testid="test" aria-label="test" />
  );
  expect(getByTestId("test")).toHaveAttribute("aria-label", "test");
});

test("fillColor", () => {
  const { container } = render(<PopoverArrow fillColor="red" />);
  expect(container.firstChild).toMatchSnapshot();
});

test("strokeColor", () => {
  const { container } = render(<PopoverArrow strokeColor="red" />);
  expect(container.firstChild).toMatchSnapshot();
});

test("styled", () => {
  const { container } = render(<PopoverArrow />);
  expect(container.firstChild).toMatchSnapshot();
});
