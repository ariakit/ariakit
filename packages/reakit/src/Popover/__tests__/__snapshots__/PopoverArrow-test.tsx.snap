// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fillColor 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  position: absolute;
  font-size: 30px;
  width: 1em;
  height: 1em;
  pointer-events: none;
  background-color: transparent;
}

.c0 .stroke {
  fill: transparent;
}

.c0 .fill {
  fill: red;
}

[data-placement^="top"] > .c0 {
  top: 100%;
  -webkit-transform: rotateZ(180deg);
  -ms-transform: rotateZ(180deg);
  transform: rotateZ(180deg);
}

[data-placement^="right"] > .c0 {
  right: 100%;
  -webkit-transform: rotateZ(-90deg);
  -ms-transform: rotateZ(-90deg);
  transform: rotateZ(-90deg);
}

[data-placement^="bottom"] > .c0 {
  bottom: 100%;
  -webkit-transform: rotateZ(360deg);
  -ms-transform: rotateZ(360deg);
  transform: rotateZ(360deg);
}

[data-placement^="left"] > .c0 {
  left: 100%;
  -webkit-transform: rotateZ(90deg);
  -ms-transform: rotateZ(90deg);
  transform: rotateZ(90deg);
}

<div
  class="c0 c1"
>
  <svg
    viewBox="0 0 30 30"
  >
    <path
      class="stroke"
      d="M23.7,27.1L17,19.9C16.5,19.3,15.8,19,15,19s-1.6,0.3-2.1,0.9l-6.6,7.2C5.3,28.1,3.4,29,2,29h26 C26.7,29,24.6,28.1,23.7,27.1z"
    />
    <path
      class="fill"
      d="M23,27.8c1.1,1.2,3.4,2.2,5,2.2h2H0h2c1.7,0,3.9-1,5-2.2l6.6-7.2c0.7-0.8,2-0.8,2.7,0L23,27.8L23,27.8z"
    />
  </svg>
</div>
`;

exports[`strokeColor 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  position: absolute;
  font-size: 30px;
  width: 1em;
  height: 1em;
  pointer-events: none;
  background-color: transparent;
}

.c0 .stroke {
  fill: red;
}

.c0 .fill {
  fill: unset;
}

[data-placement^="top"] > .c0 {
  top: 100%;
  -webkit-transform: rotateZ(180deg);
  -ms-transform: rotateZ(180deg);
  transform: rotateZ(180deg);
}

[data-placement^="right"] > .c0 {
  right: 100%;
  -webkit-transform: rotateZ(-90deg);
  -ms-transform: rotateZ(-90deg);
  transform: rotateZ(-90deg);
}

[data-placement^="bottom"] > .c0 {
  bottom: 100%;
  -webkit-transform: rotateZ(360deg);
  -ms-transform: rotateZ(360deg);
  transform: rotateZ(360deg);
}

[data-placement^="left"] > .c0 {
  left: 100%;
  -webkit-transform: rotateZ(90deg);
  -ms-transform: rotateZ(90deg);
  transform: rotateZ(90deg);
}

<div
  class="c0 c1"
>
  <svg
    viewBox="0 0 30 30"
  >
    <path
      class="stroke"
      d="M23.7,27.1L17,19.9C16.5,19.3,15.8,19,15,19s-1.6,0.3-2.1,0.9l-6.6,7.2C5.3,28.1,3.4,29,2,29h26 C26.7,29,24.6,28.1,23.7,27.1z"
    />
    <path
      class="fill"
      d="M23,27.8c1.1,1.2,3.4,2.2,5,2.2h2H0h2c1.7,0,3.9-1,5-2.2l6.6-7.2c0.7-0.8,2-0.8,2.7,0L23,27.8L23,27.8z"
    />
  </svg>
</div>
`;

exports[`styled 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  position: absolute;
  font-size: 30px;
  width: 1em;
  height: 1em;
  pointer-events: none;
  background-color: transparent;
}

.c0 .stroke {
  fill: transparent;
}

.c0 .fill {
  fill: unset;
}

[data-placement^="top"] > .c0 {
  top: 100%;
  -webkit-transform: rotateZ(180deg);
  -ms-transform: rotateZ(180deg);
  transform: rotateZ(180deg);
}

[data-placement^="right"] > .c0 {
  right: 100%;
  -webkit-transform: rotateZ(-90deg);
  -ms-transform: rotateZ(-90deg);
  transform: rotateZ(-90deg);
}

[data-placement^="bottom"] > .c0 {
  bottom: 100%;
  -webkit-transform: rotateZ(360deg);
  -ms-transform: rotateZ(360deg);
  transform: rotateZ(360deg);
}

[data-placement^="left"] > .c0 {
  left: 100%;
  -webkit-transform: rotateZ(90deg);
  -ms-transform: rotateZ(90deg);
  transform: rotateZ(90deg);
}

<div
  class="c0 c1"
>
  <svg
    viewBox="0 0 30 30"
  >
    <path
      class="stroke"
      d="M23.7,27.1L17,19.9C16.5,19.3,15.8,19,15,19s-1.6,0.3-2.1,0.9l-6.6,7.2C5.3,28.1,3.4,29,2,29h26 C26.7,29,24.6,28.1,23.7,27.1z"
    />
    <path
      class="fill"
      d="M23,27.8c1.1,1.2,3.4,2.2,5,2.2h2H0h2c1.7,0,3.9-1,5-2.2l6.6-7.2c0.7-0.8,2-0.8,2.7,0L23,27.8L23,27.8z"
    />
  </svg>
</div>
`;
